### PR TITLE
Update Client interface

### PR DESCRIPTION
**Pull request by donny-dont**
This updates the `Client` interface exposed and removes the extensions of `Request` and `Response` that are no longer applicable.

### CODE REVIEW

File: lib/http.dart [hunk 1]
```diff
@@ -32,7 +32,7 @@ export 'src/streamed_response.dart';
 /// the same server, you should use a single [Client] for all of those requests.
 ///
 /// For more fine-grained control over the request, use [Request] instead.
-Future<Response> head(url, {Map<String, String> headers}) =>
+FutureOr<Response> head(url, {Map<String, String> headers}) =>
```
Comment (kevmoo) [inline lib/http.dart]: Uh...these are *always* async, right?

Why the move to `FutureOr`?

Comment (donny-dont) [inline lib/http.dart]: @kevmoo the intent is to move to middleware like shelf.

Comment (kevmoo) [inline lib/http.dart]: Okay...but do we imagine a future where these calls may be sync?

Comment (donny-dont) [inline lib/http.dart]: Are we just talking http.dart or throughout? With the http helpers no.

Comment (kevmoo) [inline lib/http.dart]: Exactly. So this method should always return `Future`. `FutureOr` implies that it may synchronously return `Response`.

**Comment by donny-dont** [comment]
@kevmoo fixed the interface.

File: lib/http.dart [hunk 2]
```diff
@@ -161,7 +157,7 @@ Future<String> read(url, {Map<String, String> headers}) =>
 Future<Uint8List> readBytes(url, {Map<String, String> headers}) =>
   _withClient((client) => client.readBytes(url, headers: headers));
 
-Future/*<T>*/ _withClient/*<T>*/(Future/*<T>*/ fn(Client client)) async {
+Future<T> _withClient<T>(FutureOr<T> fn(Client client)) async {
```
Comment (nex3) [inline lib/http.dart]: Why is this `FutureOr`? I'm pretty sure all the functions return `Future`s.

File: lib/src/base_client.dart [hunk 3]
```diff
@@ -23,15 +21,15 @@ abstract class BaseClient implements Client {
   /// can be a [Uri] or a [String].
   ///
   /// For more fine-grained control over the request, use [send] instead.
-  Future<Response> head(url, {Map<String, String> headers}) =>
-    _sendUnstreamed("HEAD", url, headers);
+  FutureOr<Response> head(url, {Map<String, String> headers}) =>
```
Comment (nex3) [inline lib/src/base_client.dart]: Same question here: why `FutureOr`? This is never going to return a `Response` synchronously.

File: lib/src/base_client.dart [hunk 4]
```diff
@@ -23,15 +21,15 @@ abstract class BaseClient implements Client {
...
+    send(new Request.head(_uri(url), headers: headers));
```
Comment (nex3) [inline lib/src/base_client.dart]: The request constructors should take dynamic `url`s, not `Uri`s.

Comment (donny-dont) [inline lib/src/base_client.dart]: How about the named constructors take dynamic but the default takes a Uri? If all of them take dynamic then there's a check on each call to `change`.

Comment (nex3) [inline lib/src/base_client.dart]: I'd rather have a cleaner public API and add a private constructor for `change` to call, I think.

Comment (donny-dont) [inline lib/src/base_client.dart]: K works for me 👍

**Comment by donny-dont** [comment]
@nex3 @kevmoo think we're ready to go know.

File: pubspec.yaml [hunk 5]
```diff
@@ -4,7 +4,7 @@ author: "Dart Team <misc@dartlang.org>"
 homepage: https://github.com/dart-lang/http
```
Comment (kevmoo) [inline pubspec.yaml]: We should make sure to update the pubspec version to make a major change, too.

Comment (kevmoo) [inline pubspec.yaml]: We should make sure to update the pubspec version to make a major change, too.

Comment (donny-dont) [inline pubspec.yaml]: Updated to 0.12.0

Comment (kevmoo) [inline pubspec.yaml]: 👍

File: lib/src/request.dart [hunk 6]
```diff
@@ -147,10 +151,22 @@ class Request extends Message {
...
+        this.method,
+        this.url,
+        body ?? getBody(this),
+        this.encoding,
+        updatedHeaders,
+        updatedContext);
+  }
+}
+
+Uri _uri(url) {
```
Comment (nex3) [inline lib/src/request.dart]: Either make this a static method of `Request`, or put it in `utils`.

File: lib/src/request.dart [hunk 7]
```diff
@@ -30,12 +30,12 @@ class Request extends Message {
   ///
   /// Extra [context] can be used to pass information between inner middleware
   /// and handlers.
-  Request(this.method, this.url,
+  Request(String method, url,
```
Comment (nex3) [inline lib/src/request.dart]: Document the acceptable types for `url`. You can probably just copy the docs from the top-level methods.

File: pubspec.yaml (hunk 5 above)
Comment (nex3) [inline pubspec.yaml]: This should be 0.12.0-dev. We use the "-dev" suffix to clearly distinguish between released versions and those that are still in progress.

**Comment by donny-dont** [comment]
🎊

**Pull request merged**